aspnetcore/mvc/models/file-uploads.md

```diff
@@ -1,195 +1,656 @@
+* **Run a virus/malware scanner on uploaded content before the file is stored.**
+
+&dagger;The sample app demonstrates an approach that meets the criteria.
+
+> [!WARNING]
+> Uploading malicious code to a system is frequently the first step to executing code that can:
+>
+> * Completely takeover a system.
+> * Overload a system with the result that the system crashes.
+> * Compromise user or system data.
+> * Apply graffiti to a public UI.
+>
+> For information on reducing the attack surface area when accepting files from users, see the following resources:
+>
+> * [Unrestricted File Upload](https://www.owasp.org/index.php/Unrestricted_File_Upload)
+> * [Azure Security: Ensure appropriate controls are in place when accepting files from users](/azure/security/azure-security-threat-modeling-tool-input-validation#controls-users)
+
+For more information on implementing security measures, including examples from the sample app, see the [Validation](#validation) section.
+
+## Storage scenarios
+
+Common storage options for files include:
+ 
+* Database
+
+  * For small file uploads, often faster than physical storage (file system or network share) options.
+  * Often more convenient than physical storage options because retrieval of a database record for user data can concurrently supply the file content (for example, an avatar image).
+  * Potentially less expensive than using a data storage service.
+
+* Physical storage (file system or network share)
+
+  * For large file uploads:
+    * Database limits may restrict the size of the upload.
+    * Often less economical than storage in a database.
+  * Potentially less expensive than using a data storage service.
+  * The app's process must have read and write permissions to the storage location. **Never grant execute permission.**
+
+* Data storage service (for example, [Azure Blob Storage](https://azure.microsoft.com/services/storage/blobs/))
+
+  * Improved scalability and resiliency over on-premises solutions that are often subject to single points of failure.
+  * Potentially lower cost in large storage infrastructure scenarios.
+
+  For more information, see [Quickstart: Use .NET to create a blob in object storage](/azure/storage/blobs/storage-quickstart-blobs-dotnet). The topic demonstrates <xref:Microsoft.Azure.Storage.File.CloudFile.UploadFromFileAsync*>, but you can use <xref:Microsoft.Azure.Storage.File.CloudFile.UploadFromStreamAsync*> to save a <xref:System.IO.FileStream> to blob storage when working with a <xref:System.IO.Stream>.
+
+## File upload scenarios
+
+Two general approaches for uploading files are buffering and streaming.
+
+**Buffering**
+
+The entire file is read into an <xref:Microsoft.AspNetCore.Http.IFormFile>, which is a C# representation of the file used to process or save the file.
+
+> [!NOTE]
+> Any single buffered file exceeding 64 KB is moved from memory to a temp file on disk.
```

Comment:
Should we call out that the temp file is opaque to the user and entirely managed by ASP.NET Core? Also is this useful for users to know? Were users concerned about buffering large files in memory?

Reply:
Generally, I'd say 'yes' to that. It makes one aware, for example, that an app working with files just a hair over 64KB will incur an immediate perf penalty over one that always handles files <64 KB.

I'll move it down in the section a paragraph to de-emphasize it a bit.

Reply:
@pranavkm To answer the question "Is it useful for the user to know?" Yes it is. As an user we have had problems with this in the past because we did not know this happens. For example we gracefully wanted to handle the error flow when there is a virus but if you do not know it writes to disk the virusscanner will react and throw it away.